From e6e00eb355382a1a39d36acf56973ae9330ff836 Mon Sep 17 00:00:00 2001
From: yuanhecai <yuanhecai@loongson.cn>
Date: Thu, 13 Jun 2024 11:12:16 +0800
Subject: [PATCH 15/15] FROMLIST: LoongArch64: Optimize functions in
 blockcopy8.S file - Patch 2

The optimized functions are as follows:
x265_cpy2Dto1D_shl*_lsx/_lasx
x265_cpy2Dto1D_shr*_lsx/_lasx
x265_cpy1Dto2D_shl*_lsx/_lasx
x265_cpy1Dto2D_shr*_lsx/_lasx

Link: https://bitbucket.org/multicoreware/x265_git/pull-requests/31
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 source/common/loongarch64/asm-primitives.cpp |  36 +-
 source/common/loongarch64/blockcopy8.S       | 580 +++++++++++++++++++
 source/common/loongarch64/blockcopy8.h       |  28 +
 3 files changed, 643 insertions(+), 1 deletion(-)

diff --git a/source/common/loongarch64/asm-primitives.cpp b/source/common/loongarch64/asm-primitives.cpp
index 197e8df96..6f88d2bd5 100644
--- a/source/common/loongarch64/asm-primitives.cpp
+++ b/source/common/loongarch64/asm-primitives.cpp
@@ -1,5 +1,5 @@
 /*****************************************************************************
- * Copyright (C) 2013-2024 MulticoreWare, Inc
+ * Copyright (C) 2024 MulticoreWare, Inc
  *
  * Authors: Hao Chen <chenhao@loongson.cn>
  *
@@ -257,6 +257,11 @@ all_angs_pred_lsx(32, 10)
     p.cu[BLOCK_ ## W ## x ## H].copy_ss     = PFX(blockcopy_ss_ ## W ## x ## H ##_ ## CPU); \
     p.cu[BLOCK_ ## W ## x ## H].copy_sp     = PFX(blockcopy_sp_ ## W ## x ## H ##_ ## CPU); \
     p.cu[BLOCK_ ## W ## x ## H].copy_ps     = PFX(blockcopy_ps_ ## W ## x ## H ##_ ## CPU); \
+    p.cu[BLOCK_ ## W ## x ## H].cpy2Dto1D_shl = PFX(cpy2Dto1D_shl_ ## W ## x ## H ##_ ## CPU); \
+    p.cu[BLOCK_ ## W ## x ## H].cpy2Dto1D_shr = PFX(cpy2Dto1D_shr_ ## W ## x ## H ##_ ## CPU); \
+    p.cu[BLOCK_ ## W ## x ## H].cpy1Dto2D_shl[NONALIGNED] = PFX(cpy1Dto2D_shl_ ## W ## x ## H ##_ ## CPU); \
+    p.cu[BLOCK_ ## W ## x ## H].cpy1Dto2D_shl[ALIGNED] = PFX(cpy1Dto2D_shl_ ## W ## x ## H ##_ ## CPU); \
+    p.cu[BLOCK_ ## W ## x ## H].cpy1Dto2D_shr = PFX(cpy1Dto2D_shr_ ## W ## x ## H ##_ ## CPU); \
     p.cu[BLOCK_ ## W ## x ## H].transpose     = PFX(transpose_ ## W ## x ## H ##_ ## CPU); \
     p.cu[BLOCK_ ## W ## x ## H].calcresidual[NONALIGNED] = PFX(pixel_getResidual_ ## W ## x ## H ##_ ## CPU); \
     p.cu[BLOCK_ ## W ## x ## H].calcresidual[ALIGNED]    = PFX(pixel_getResidual_ ## W ## x ## H ##_ ## CPU);;
@@ -1392,6 +1397,35 @@ void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
         //copy_cnt
         p.cu[BLOCK_16x16].copy_cnt = PFX(copy_count_16_lasx);
         p.cu[BLOCK_32x32].copy_cnt = PFX(copy_count_32_lasx);
+
+        // cpy2Dto1D_shl
+        p.cu[BLOCK_8x8].cpy2Dto1D_shl   = PFX(cpy2Dto1D_shl_8x8_lasx);
+        p.cu[BLOCK_16x16].cpy2Dto1D_shl = PFX(cpy2Dto1D_shl_16x16_lasx);
+        p.cu[BLOCK_32x32].cpy2Dto1D_shl = PFX(cpy2Dto1D_shl_32x32_lasx);
+        p.cu[BLOCK_64x64].cpy2Dto1D_shl = PFX(cpy2Dto1D_shl_64x64_lasx);
+
+        // cpy2Dto1D_shr
+        p.cu[BLOCK_8x8].cpy2Dto1D_shr   = PFX(cpy2Dto1D_shr_8x8_lasx);
+        p.cu[BLOCK_16x16].cpy2Dto1D_shr = PFX(cpy2Dto1D_shr_16x16_lasx);
+        p.cu[BLOCK_32x32].cpy2Dto1D_shr = PFX(cpy2Dto1D_shr_32x32_lasx);
+        p.cu[BLOCK_64x64].cpy2Dto1D_shr = PFX(cpy2Dto1D_shr_64x64_lasx);
+
+        // cpy1Dto2D_shl
+        p.cu[BLOCK_8x8].cpy1Dto2D_shl[NONALIGNED]   = PFX(cpy1Dto2D_shl_8x8_lasx);
+        p.cu[BLOCK_16x16].cpy1Dto2D_shl[NONALIGNED] = PFX(cpy1Dto2D_shl_16x16_lasx);
+        p.cu[BLOCK_32x32].cpy1Dto2D_shl[NONALIGNED] = PFX(cpy1Dto2D_shl_32x32_lasx);
+        p.cu[BLOCK_64x64].cpy1Dto2D_shl[NONALIGNED] = PFX(cpy1Dto2D_shl_64x64_lasx);
+
+        p.cu[BLOCK_8x8].cpy1Dto2D_shl[ALIGNED]   = PFX(cpy1Dto2D_shl_8x8_lasx);
+        p.cu[BLOCK_16x16].cpy1Dto2D_shl[ALIGNED] = PFX(cpy1Dto2D_shl_16x16_lasx);
+        p.cu[BLOCK_32x32].cpy1Dto2D_shl[ALIGNED] = PFX(cpy1Dto2D_shl_32x32_lasx);
+        p.cu[BLOCK_64x64].cpy1Dto2D_shl[ALIGNED] = PFX(cpy1Dto2D_shl_64x64_lasx);
+
+        // cpy1Dto2D_shr
+        p.cu[BLOCK_8x8].cpy1Dto2D_shr   = PFX(cpy1Dto2D_shr_8x8_lasx);
+        p.cu[BLOCK_16x16].cpy1Dto2D_shr = PFX(cpy1Dto2D_shr_16x16_lasx);
+        p.cu[BLOCK_32x32].cpy1Dto2D_shr = PFX(cpy1Dto2D_shr_32x32_lasx);
+        p.cu[BLOCK_64x64].cpy1Dto2D_shr = PFX(cpy1Dto2D_shr_64x64_lasx);
     }
 #endif /* HAVE_LASX */
 }
diff --git a/source/common/loongarch64/blockcopy8.S b/source/common/loongarch64/blockcopy8.S
index 475590abe..42752a7c4 100644
--- a/source/common/loongarch64/blockcopy8.S
+++ b/source/common/loongarch64/blockcopy8.S
@@ -2047,3 +2047,583 @@ function x265_copy_count_32_lasx
     bnez          t3,     1b
     or            a0,     zero,   t2
 endfunc
+
+function x265_cpy2Dto1D_shr_4x4_lsx
+    slli.d         a2,      a2,    1
+    fld.d          f0,      a1,    0
+    fldx.d         f1,      a1,    a2
+    vreplgr2vr.h   vr3,     a3
+    vilvl.d        vr2,     vr1,   vr0
+    vsrar.h        vr4,     vr2,   vr3
+    vst            vr4,     a0,    0
+
+    alsl.d         a1,      a2,    a1,   1
+
+    fld.d          f0,      a1,    0
+    fldx.d         f1,      a1,    a2
+    vreplgr2vr.h   vr3,     a3
+    vilvl.d        vr2,     vr1,   vr0
+    vsrar.h        vr4,     vr2,   vr3
+    vst            vr4,     a0,    16
+endfunc
+
+function x265_cpy2Dto1D_shr_8x8_lsx
+    slli.d         a2,      a2,    1
+.rept 4
+    vld            vr0,     a1,    0
+    vldx           vr1,     a1,    a2
+    vreplgr2vr.h   vr3,     a3
+    vsrar.h        vr4,     vr0,   vr3
+    vsrar.h        vr5,     vr1,   vr3
+    vst            vr4,     a0,    0
+    vst            vr5,     a0,    16
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    32
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_16x16_lsx
+    slli.d         a2,      a2,    1
+.rept 8
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    add.d          t1,      a1,    a2
+    vld            vr2,     t1,    0
+    vld            vr3,     t1,    16
+    vreplgr2vr.h   vr4,     a3
+    vsrar.h        vr5,     vr0,   vr4
+    vsrar.h        vr6,     vr1,   vr4
+    vsrar.h        vr7,     vr2,   vr4
+    vsrar.h        vr8,     vr3,   vr4
+    vst            vr5,     a0,    0
+    vst            vr6,     a0,    16
+    vst            vr7,     a0,    32
+    vst            vr8,     a0,    48
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    64
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_32x32_lsx
+    slli.d         a2,      a2,    1
+.rept 16
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr10,    a1,    32
+    vld            vr11,    a1,    48
+    add.d          t1,      a1,    a2
+    vld            vr2,     t1,    0
+    vld            vr3,     t1,    16
+    vld            vr12,    t1,    32
+    vld            vr13,    t1,    48
+    vreplgr2vr.h   vr4,     a3
+
+.irp i, vr0, vr1, vr10, vr11, vr2, vr3, vr12, vr13
+    vsrar.h        \i,      \i,   vr4
+.endr
+
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr10,    a0,    32
+    vst            vr11,    a0,    48
+    vst            vr2,     a0,    64
+    vst            vr3,     a0,    80
+    vst            vr12,    a0,    96
+    vst            vr13,    a0,    112
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    128
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_64x64_lsx
+    slli.d         a2,      a2,    1
+.rept 64
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr2,     a1,    32
+    vld            vr3,     a1,    48
+    vld            vr4,     a1,    64
+    vld            vr5,     a1,    80
+    vld            vr6,     a1,    96
+    vld            vr7,     a1,    112
+    vreplgr2vr.h   vr8,     a3
+
+.irp i, vr0, vr1, vr2, vr3, vr4, vr5, vr6, vr7
+    vsrar.h        \i,      \i,    vr8
+.endr
+
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr2,     a0,    32
+    vst            vr3,     a0,    48
+    vst            vr4,     a0,    64
+    vst            vr5,     a0,    80
+    vst            vr6,     a0,    96
+    vst            vr7,     a0,    112
+    add.d          a1,      a2,    a1
+    addi.d         a0,      a0,    128
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_8x8_lasx
+    slli.d         a2,      a2,    1
+.rept 4
+    vld            vr0,     a1,    0
+    vldx           vr1,     a1,    a2
+    xvpermi.q      xr1,     xr0,   0x20
+    xvreplgr2vr.h  xr3,     a3
+    xvsrar.h       xr4,     xr1,   xr3
+    xvst           xr4,     a0,    0
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    32
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_16x16_lasx
+    slli.d         a2,      a2,    1
+.rept 8
+    xvld           xr0,     a1,    0
+    xvldx          xr1,     a1,    a2
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr6,     xr0,   xr4
+    xvsrar.h       xr8,     xr1,   xr4
+    xvst           xr6,     a0,    0
+    xvst           xr8,     a0,    32
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    64
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_32x32_lasx
+    slli.d         a2,      a2,    1
+.rept 16
+    xvld           xr0,     a1,    0
+    xvldx          xr1,     a1,    a2
+    addi.d         t1,      a1,    32
+    xvld           xr2,     t1,    0
+    xvldx          xr3,     t1,    a2
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr6,     xr0,   xr4
+    xvsrar.h       xr7,     xr2,   xr4
+    xvsrar.h       xr8,     xr1,   xr4
+    xvsrar.h       xr9,     xr3,   xr4
+    xvst           xr6,     a0,    0
+    xvst           xr7,     a0,    32
+    xvst           xr8,     a0,    64
+    xvst           xr9,     a0,    96
+    alsl.d         a1,      a2,    a1,   1
+    addi.d         a0,      a0,    128
+.endr
+endfunc
+
+function x265_cpy2Dto1D_shr_64x64_lasx
+    slli.d         a2,      a2,    1
+.rept 64
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr6,     xr0,   xr4
+    xvsrar.h       xr7,     xr2,   xr4
+    xvsrar.h       xr8,     xr1,   xr4
+    xvsrar.h       xr9,     xr3,   xr4
+    xvst           xr6,     a0,    0
+    xvst           xr7,     a0,    32
+    xvst           xr8,     a0,    64
+    xvst           xr9,     a0,    96
+    add.d          a1,      a2,    a1
+    addi.d         a0,      a0,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_4x4_lsx
+    slli.d         a2,      a2,    1
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vreplgr2vr.h   vr3,     a3
+    vsll.h         vr4,     vr0,   vr3
+    vsll.h         vr5,     vr1,   vr3
+    vstelm.d       vr4,     a0,    0,    0
+    add.d          a0,      a0,    a2
+    vstelm.d       vr4,     a0,    0,    1
+    add.d          a0,      a0,    a2
+    vstelm.d       vr5,     a0,    0,    0
+    add.d          a0,      a0,    a2
+    vstelm.d       vr5,     a0,    0,    1
+endfunc
+
+function x265_cpy1Dto2D_shl_8x8_lsx
+    slli.d         a2,      a2,    1
+.rept 4
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vreplgr2vr.h   vr3,     a3
+    vsll.h         vr4,     vr0,   vr3
+    vsll.h         vr5,     vr1,   vr3
+    vst            vr4,     a0,    0
+    vstx           vr5,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    addi.d         a1,      a1,    32
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_16x16_lsx
+    slli.d         a2,      a2,    1
+.rept 8
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr2,     a1,    32
+    vld            vr3,     a1,    48
+    vreplgr2vr.h   vr4,     a3
+    vsll.h         vr5,     vr0,   vr4
+    vsll.h         vr6,     vr1,   vr4
+    vsll.h         vr7,     vr2,   vr4
+    vsll.h         vr8,     vr3,   vr4
+    vst            vr5,     a0,    0
+    vst            vr6,     a0,    16
+    add.d          a0,      a2,    a0
+    vst            vr7,     a0,    0
+    vst            vr8,     a0,    16
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    64
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_32x32_lsx
+    slli.d         a2,      a2,    1
+.rept 16
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr10,    a1,    32
+    vld            vr11,    a1,    48
+    vld            vr2,     a1,    64
+    vld            vr3,     a1,    80
+    vld            vr12,    a1,    96
+    vld            vr13,    a1,    112
+    vreplgr2vr.h   vr4,     a3
+.irp i, vr0, vr1, vr10, vr11, vr2, vr3, vr12, vr13
+    vsll.h         \i,      \i,   vr4
+.endr
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr10,    a0,    32
+    vst            vr11,    a0,    48
+    add.d          a0,      a0,    a2
+    vst            vr2,     a0,    0
+    vst            vr3,     a0,    16
+    vst            vr12,    a0,    32
+    vst            vr13,    a0,    48
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_64x64_lsx
+    slli.d         a2,      a2,    1
+.rept 64
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr2,     a1,    32
+    vld            vr3,     a1,    48
+    vld            vr4,     a1,    64
+    vld            vr5,     a1,    80
+    vld            vr6,     a1,    96
+    vld            vr7,     a1,    112
+    vreplgr2vr.h   vr8,     a3
+.irp i, vr0, vr1, vr2, vr3, vr4, vr5, vr6, vr7
+    vsll.h         \i,      \i,    vr8
+.endr
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr2,     a0,    32
+    vst            vr3,     a0,    48
+    vst            vr4,     a0,    64
+    vst            vr5,     a0,    80
+    vst            vr6,     a0,    96
+    vst            vr7,     a0,    112
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_8x8_lasx
+    slli.d         a2,      a2,    1
+
+.rept 2
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvreplgr2vr.h  xr3,     a3
+    xvsll.h        xr4,     xr0,   xr3
+    xvsll.h        xr5,     xr1,   xr3
+    vst            vr4,     a0,    0
+    xvpermi.q      xr6,     xr4,   0x01
+    vstx           vr6,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    vst            vr5,     a0,    0
+    xvpermi.q      xr6,     xr5,   0x01
+    vstx           vr6,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    addi.d         a1,      a1,    64
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_16x16_lasx
+    slli.d         a2,      a2,    1
+.rept 4
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsll.h        xr5,     xr0,   xr4
+    xvsll.h        xr6,     xr1,   xr4
+    xvsll.h        xr7,     xr2,   xr4
+    xvsll.h        xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvstx          xr6,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    xvst           xr7,     a0,    0
+    xvstx          xr8,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_32x32_lasx
+    slli.d         a2,      a2,    1
+.rept 16
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsll.h        xr5,     xr0,   xr4
+    xvsll.h        xr6,     xr1,   xr4
+    xvsll.h        xr7,     xr2,   xr4
+    xvsll.h        xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvst           xr6,     a0,    32
+    add.d          a0,      a2,    a0
+    xvst           xr7,     a0,    0
+    xvst           xr8,     a0,    32
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shl_64x64_lasx
+    slli.d         a2,      a2,    1
+.rept 64
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsll.h        xr5,     xr0,   xr4
+    xvsll.h        xr6,     xr1,   xr4
+    xvsll.h        xr7,     xr2,   xr4
+    xvsll.h        xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvst           xr6,     a0,    32
+    xvst           xr7,     a0,    64
+    xvst           xr8,     a0,    96
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_4x4_lsx
+    slli.d         a2,      a2,    1
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vreplgr2vr.h   vr3,     a3
+    vsrar.h        vr4,     vr0,   vr3
+    vsrar.h        vr5,     vr1,   vr3
+    vstelm.d       vr4,     a0,    0,    0
+    add.d          a0,      a0,    a2
+    vstelm.d       vr4,     a0,    0,    1
+    add.d          a0,      a0,    a2
+    vstelm.d       vr5,     a0,    0,    0
+    add.d          a0,      a0,    a2
+    vstelm.d       vr5,     a0,    0,    1
+endfunc
+
+function x265_cpy1Dto2D_shr_8x8_lsx
+    slli.d         a2,      a2,    1
+.rept 4
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vreplgr2vr.h   vr3,     a3
+    vsrar.h        vr4,     vr0,   vr3
+    vsrar.h        vr5,     vr1,   vr3
+    vst            vr4,     a0,    0
+    vstx           vr5,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    addi.d         a1,      a1,    32
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_16x16_lsx
+    slli.d         a2,      a2,    1
+.rept 8
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr2,     a1,    32
+    vld            vr3,     a1,    48
+    vreplgr2vr.h   vr4,     a3
+    vsrar.h        vr5,     vr0,   vr4
+    vsrar.h        vr6,     vr1,   vr4
+    vsrar.h        vr7,     vr2,   vr4
+    vsrar.h        vr8,     vr3,   vr4
+    vst            vr5,     a0,    0
+    vst            vr6,     a0,    16
+    add.d          a0,      a2,    a0
+    vst            vr7,     a0,    0
+    vst            vr8,     a0,    16
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    64
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_32x32_lsx
+    slli.d         a2,      a2,    1
+.rept 16
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr10,    a1,    32
+    vld            vr11,    a1,    48
+    vld            vr2,     a1,    64
+    vld            vr3,     a1,    80
+    vld            vr12,    a1,    96
+    vld            vr13,    a1,    112
+    vreplgr2vr.h   vr4,     a3
+.irp i, vr0, vr1, vr10, vr11, vr2, vr3, vr12, vr13
+    vsrar.h        \i,      \i,   vr4
+.endr
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr10,    a0,    32
+    vst            vr11,    a0,    48
+    add.d          a0,      a0,    a2
+    vst            vr2,     a0,    0
+    vst            vr3,     a0,    16
+    vst            vr12,    a0,    32
+    vst            vr13,    a0,    48
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_64x64_lsx
+    slli.d         a2,      a2,    1
+.rept 64
+    vld            vr0,     a1,    0
+    vld            vr1,     a1,    16
+    vld            vr2,     a1,    32
+    vld            vr3,     a1,    48
+    vld            vr4,     a1,    64
+    vld            vr5,     a1,    80
+    vld            vr6,     a1,    96
+    vld            vr7,     a1,    112
+    vreplgr2vr.h   vr8,     a3
+.irp i, vr0, vr1, vr2, vr3, vr4, vr5, vr6, vr7
+    vsrar.h        \i,      \i,    vr8
+.endr
+    vst            vr0,     a0,    0
+    vst            vr1,     a0,    16
+    vst            vr2,     a0,    32
+    vst            vr3,     a0,    48
+    vst            vr4,     a0,    64
+    vst            vr5,     a0,    80
+    vst            vr6,     a0,    96
+    vst            vr7,     a0,    112
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_8x8_lasx
+    slli.d         a2,      a2,    1
+.rept 2
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvreplgr2vr.h  xr3,     a3
+    xvsrar.h       xr4,     xr0,   xr3
+    xvsrar.h       xr5,     xr1,   xr3
+    xvpermi.q      xr6,     xr4,   0x01
+    xvpermi.q      xr7,     xr5,   0x01
+    vst            vr4,     a0,    0
+    vstx           vr6,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    vst            vr5,     a0,    0
+    vstx           vr7,     a0,    a2
+    addi.d         a1,      a1,    64
+    alsl.d         a0,      a2,    a0,   1
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_16x16_lasx
+    slli.d         a2,      a2,    1
+.rept 4
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr5,     xr0,   xr4
+    xvsrar.h       xr6,     xr1,   xr4
+    xvsrar.h       xr7,     xr2,   xr4
+    xvsrar.h       xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvstx          xr6,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    xvst           xr7,     a0,    0
+    xvstx          xr8,     a0,    a2
+    alsl.d         a0,      a2,    a0,   1
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_32x32_lasx
+    slli.d         a2,      a2,    1
+.rept 16
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr5,     xr0,   xr4
+    xvsrar.h       xr6,     xr1,   xr4
+    xvsrar.h       xr7,     xr2,   xr4
+    xvsrar.h       xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvst           xr6,     a0,    32
+    add.d          a0,      a2,    a0
+    xvst           xr7,     a0,    0
+    xvst           xr8,     a0,    32
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
+
+function x265_cpy1Dto2D_shr_64x64_lasx
+    slli.d         a2,      a2,    1
+.rept 64
+    xvld           xr0,     a1,    0
+    xvld           xr1,     a1,    32
+    xvld           xr2,     a1,    64
+    xvld           xr3,     a1,    96
+    xvreplgr2vr.h  xr4,     a3
+    xvsrar.h       xr5,     xr0,   xr4
+    xvsrar.h       xr6,     xr1,   xr4
+    xvsrar.h       xr7,     xr2,   xr4
+    xvsrar.h       xr8,     xr3,   xr4
+    xvst           xr5,     a0,    0
+    xvst           xr6,     a0,    32
+    xvst           xr7,     a0,    64
+    xvst           xr8,     a0,    96
+    add.d          a0,      a2,    a0
+    addi.d         a1,      a1,    128
+.endr
+endfunc
diff --git a/source/common/loongarch64/blockcopy8.h b/source/common/loongarch64/blockcopy8.h
index f1583189f..c3ce832b5 100644
--- a/source/common/loongarch64/blockcopy8.h
+++ b/source/common/loongarch64/blockcopy8.h
@@ -168,6 +168,34 @@ void x265_blockcopy_ps_64x64_lasx(int16_t* a, intptr_t stridea, const pixel* b,
 uint32_t x265_copy_count_16_lasx(int16_t* coeff, const int16_t* residual, intptr_t resiStride);
 uint32_t x265_copy_count_32_lasx(int16_t* coeff, const int16_t* residual, intptr_t resiStride);
 
+FUNCDEF_TU(void, cpy2Dto1D_shl, lsx, int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+
+void x265_cpy2Dto1D_shl_8x8_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shl_16x16_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shl_32x32_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shl_64x64_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+
+FUNCDEF_TU(void, cpy2Dto1D_shr, lsx, int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+
+void x265_cpy2Dto1D_shr_8x8_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shr_16x16_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shr_32x32_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+void x265_cpy2Dto1D_shr_64x64_lasx(int16_t *dst, const int16_t *src, intptr_t srcStride, int shift);
+
+FUNCDEF_TU(void, cpy1Dto2D_shl, lsx, int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+
+void x265_cpy1Dto2D_shl_8x8_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shl_16x16_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shl_32x32_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shl_64x64_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+
+FUNCDEF_TU(void, cpy1Dto2D_shr, lsx, int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+
+void x265_cpy1Dto2D_shr_8x8_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shr_16x16_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shr_32x32_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+void x265_cpy1Dto2D_shr_64x64_lasx(int16_t *dst, const int16_t *src, intptr_t dstStride, int shift);
+
 #undef FUNCDEF_PU
 #undef FUNCDEF_CHROMA_PU
 
-- 
2.52.0

